File: Backup Assistant/Core/FileListing.vb
Namespace Core

    Friend Enum BackupAction As Byte
        None
        Copy
        Delete
        Overwrite
    End Enum

    Friend Class FileListing

        Public Property IsInSource As Boolean
        Public Property IsInDestination As Boolean
        Public Property SourceLastModified As Date
        Public Property DestinationLastModified As Date
        Public Property Size As Long

        Public Sub New()
            Me.IsInSource = False
            Me.IsInDestination = False
            Me.SourceLastModified = Date.MinValue
            Me.DestinationLastModified = Date.MinValue
            Me.Size = 0L
        End Sub

        Public Function GetBackupAction() As BackupAction
            If Not IsInSource AndAlso Not IsInDestination Then
                Return BackupAction.None
            ElseIf IsInSource AndAlso Not IsInDestination Then
                Return BackupAction.Copy
            ElseIf Not IsInSource AndAlso IsInDestination Then
                Return BackupAction.Delete
            ElseIf IsInSource AndAlso IsInDestination AndAlso SourceLastModified > DestinationLastModified Then
                Return BackupAction.Overwrite
            Else ' IsInSource AndAlso IsInDestination AndAlso SourceLastModified <= DestinationLastModified
                Return BackupAction.None
            End If
        End Function

    End Class

End Namespace

File: src/DataModels/FileListing.cs
using System;

namespace BackupAssistant.DataModels
{
    public enum BackupAction : byte
    {
        None,
        Copy,
        Delete,
        Overwrite
    }

    public class FileListing
    {
        public bool IsInSource { get; set; }
        public bool IsInDestination { get; set; }
        public DateTime SourceLastModified { get; set; }
        public DateTime DestinationLastModified { get; set; }
        public long Size { get; set; }

        public FileListing()
        {
            this.IsInSource = false;
            this.IsInDestination = false;
            this.SourceLastModified = DateTime.MinValue;
            this.DestinationLastModified = DateTime.MinValue;
            this.Size = 0L;
        }

        public BackupAction GetBackupAction()
        {
            return (this.IsInSource, this.IsInDestination) switch
            {
                (false, false) => BackupAction.None,
                (true, false) => BackupAction.Copy,
                (false, true) => BackupAction.Delete,
                (true, true) when this.SourceLastModified > this.DestinationLastModified => BackupAction.Overwrite,
                (true, true) => BackupAction.None,
            };
        }
    }
}
